src/generators/non_gradient_based/feature_tweak/feature_tweak.jl

```diff
@@ -7,73 +7,77 @@ mutable struct FeatureTweakGenerator <: AbstractNonGradientBasedGenerator
 
 # Arguments
-- `generator::FeatureTweakGenerator`: The feature tweak generator.
-- `M::Models.TreeModel`: The model for which the counterfactual is generated. Must be a tree-based model.
-- `x::AbstractArray`: The factual instance.
-- `target::RawTargetType`: The target class.
+- `ce::AbstractCounterfactualExplanation`: The counterfactual explanation object.
 
 # Returns
-- `x_out::AbstractArray`: The counterfactual instance.
+- `ce::AbstractCounterfactualExplanation`: The counterfactual explanation object.
 
 # Example
-x = feature_tweaking(generator, M, x, target) # returns a counterfactual instance of `x` based on the ensemble of classifiers provided
+ce = feature_tweaking(ce) # returns a counterfactual inside the ce.s′ field based on the ensemble of classifiers provided
 """
-function feature_tweaking(
-    generator::FeatureTweakGenerator,
-    M::Models.TreeModel,
-    x::AbstractArray,
-    target::RawTargetType,
-)
-    if Models.predict_label(M, x) == target
+function feature_tweaking(ce::AbstractCounterfactualExplanation)
+    @assert isa(ce.generator, Generators.FeatureTweakGenerator) "The feature tweak algorithm can only be applied using the feature tweak generator"
+    @assert isa(ce.M, Models.TreeModel) "The `FeatureTweakGenerator` currently only supports tree models. The counterfactual search will be terminated."
+
+    if Models.predict_label(ce.M, ce.x) == ce.target
         return ce
     end
 
-    x_out = deepcopy(x)
     delta = 10^3
-    ensemble_prediction = Models.predict_label(M, x)[1]
+    ensemble_prediction = Models.predict_label(ce.M, ce.x)[1]
 
-    for classifier in Models.get_individual_classifiers(M)
-        if ensemble_prediction != target
+    for classifier in Models.get_individual_classifiers(ce.M)
+        if ensemble_prediction != ce.target
             y_levels = MLJBase.classes(
-                MLJBase.predict(M.model, DataFrames.DataFrame(x', :auto))
+                MLJBase.predict(ce.M.model, DataFrames.DataFrame(ce.x', :auto))
             )
-            paths = search_path(classifier, y_levels, target)
+            paths = search_path(classifier, y_levels, ce.target)
+
             for key in keys(paths)
                 path = paths[key]
-                es_instance = esatisfactory_instance(generator, x, path)
-                if target .== Models.predict_label(M, es_instance)[1]
-                    if LinearAlgebra.norm(x - es_instance) < delta
-                        x_out = es_instance
-                        delta = LinearAlgebra.norm(x - es_instance)
+                es_instance = esatisfactory_instance(ce.generator, ce.x, path)
+                if ce.target .== Models.predict_label(ce.M, es_instance)[1]
+                    s′_old = ce.s′
+                    ce.s′ = reshape(es_instance, :, 1)
+                    if ce.generator.penalty(ce) < delta
+                        delta = ce.generator.penalty(ce)
```

Comment:
Would it make sense to also allow for adding multiple penalties (i.e. a vector of penalty functions as for the composable generators)? Then `FeatureTweak` would fit right into the composable framework.

Reply:
Oh right, forgot about this. I think I can add that, will check whether it works out.